rfcs/20230110-kernel-reuse/README.md

```diff
@@ -0,0 +1,271 @@
+generators. Kernel generators are expected to be performant. For OpenCL C
+kernels, the generator is expected to be a program binary.
+
+#### Pros:
+Requires no user intervention.
+
+Implementation for Global primitive cache is similar.
+
+Users can control memory usage
+
+Generators can be specialized for each GPU, potentially enabling faster
+performance.
+
+#### Cons:
+Most complicated implementation (except potentially Option 2) among all the
+options.
+
+#### Implementation
+The API change will be the same as in the [global primitive
+cache](https://github.com/oneapi-src/oneDNN/tree/rfcs/rfcs/20200207-global-primitive-cache),
+except the "primitive" name will be replaced with "kernel". Kernel which use
+this cache will need to opt in to its uses. Kernels need to be designed for
+reuse for the cache to be effective.
+
+### Recommendation
+Short Answer: Start with Reuse Option 3 with a transition to a combination of
+Reuse Option 2 and Reuse Option 4 on an as need basis.
+
+There are significant benefits to modifying our implementations to generate a
+limited and finite number of kernel binaries. Among the benefits are:
+* Enables Build Time Compilation
+* Enables Runtime Parameters
+* Reduces Testing Complexity
+
+While it is impossible to achieve this in general, for sufficiently large
+problems, there tend to only be a few important compiled configurations. Because
+of these advantages of this, I recommend we start by curating a set of kernel
+binaries which address most sufficiently large problems.
+
+To reuse, I recommend starting with Reuse Option 3, the fixed size function
+registry. The motivating reason for this choice is the kernels generated for
+Reuse Option 3 also satisfy the requirements for Distribution Option 1, enabling
+to address performance issues with first compilation. In addition, this option
+requires no API changes while allowing users benefit from our improvements.
+
+In cases where we find Reuse Option 3 is insufficient, we will transition to a
+combination of Reuse Options 2 and 4 on an as need basis. After the initial work
+is performed to (as much as possible) support Reuse Option 3, transitioning to
+Reuse Option 4 will not be difficult. At an implementation level, a global
+primitive cache is very similar to a function registry. In addition, this
+curated set of kernels can be used to easily implement Reuse Option 2 which
+oneDNN users have expressed interest in. Option 1 is not being considered as it
+only handles OpenCL C kernels, which is insufficient for our needs.
+
+As a data point, a combination of Reuse Options 2 and 4 is exactly how the GEMM
+implementation works today. There are a finite number of curated GEMM
+implementations, but there are too many to store in a simple function registry
+though, so within oneMKL, these specializations are stored in a global kernel
+cache. In addition, the recommended design is part of what enabled matrix
+multiplication to support runtime dimensions in oneDNN.
```

Comment:
Quick comment/clarification: while GEMM kernels accept runtime dimensions, oneDNN GPU GEMM doesn't yet support runtime dimensions on the oneDNN API level (i.e. if users request runtime dimensions, it will go to a reference implementation).